bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,926 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.qbus.internal.protocol;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.qbus.internal.QbusBridgeHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link QbusCommunication} class is able to do the following tasks with Qbus
+ * systems:
+ * <ul>
+ * <li>Start and stop TCP socket connection with Qbus Server.
+ * <li>Read all the outputs and their status information from the Qbus Controller.
+ * <li>Execute Qbus commands.
+ * <li>Listen to events from Qbus.
+ * </ul>
+ *
+ * A class instance is instantiated from the {@link QbusBridgeHandler} class initialization.
+ *
+ * @author Koen Schockaert - Initial Contribution
+ */
+
+@NonNullByDefault
+public final class QbusCommunication {
+
+    private final Logger logger = LoggerFactory.getLogger(QbusCommunication.class);
+
+    @Nullable
+    private Socket qSocket;
+    @Nullable
+    private PrintWriter qOut;
+    @Nullable
+    private BufferedReader qIn;
+
+    private boolean listenerStopped;
+    private boolean qEventsRunning;
+
+    private Gson gsonOut = new Gson();
+    private Gson gsonIn;
+    @Nullable
+    private String CTD;
+
+    private Boolean CTDConnected = false;
```

Comment:
Primitive types should be used where possible.
```suggestion
    private boolean CTDConnected = false;
```